src/generators/notebook/set_collapsed_state.ts

```diff
@@ -0,0 +1,157 @@
+  }
+  // Guard against attempting to collapse already hidden cells...
+  if (state) {
+    if (cell.isHidden) {
+      return;
+    }
+  }
+  // Guard against attempting to expand cells which we did not collapse or are already expanded...
+  else if (
+    cell.model.metadata.has('toc-nb-collapsed') === false ||
+    cell.model.metadata.get('toc-nb-collapsed') === false
+  ) {
+    // FIXME: we can end up here when attempting to expand a section generated by a code cell
+    return;
+  }
+  if (cell.model.type !== 'markdown') {
+    // FIXME: if we're here, it's because the user is attempting to collapse/expand a section generated by a code cell
+    return;
+  }
+  const heading = parseHeading(cell.model.value.text);
+
+  // Guard against attempting to (un-)collapse cells which are not "collapsible" (i.e., do not define sections)...
+  if (heading === null) {
+    return;
+  }
+  const level = heading.level;
+
+  const widgets = tracker.currentWidget.content.widgets;
+  const len = widgets.length;
+  const idx = widgets.indexOf(cell);
+
+  // Guard against an unrecognized "cell" argument...
+  if (idx === -1) {
+    return;
+  }
+  // Search for notebook cells which are semantically defined as sub-cells...
+  for (let i = idx + 1; i < len; i++) {
+    const w = widgets[i];
+
+    // Cells which are neither Markdown nor code cells can be readily collapsed/expanded...
+    if (w.model.type !== 'markdown' && w.model.type !== 'code') {
+      w.setHidden(state);
+      continue;
+    }
+    // Markdown cells are relatively straightforward, as we can determine whether to collapse/expand based on the level of the **first** encountered heading...
+    if (w.model.type === 'markdown') {
+      const h = parseHeading(w.model.value.text);
+
+      // Check if a widget is at the same or higher level...
+      if (h && h.level >= 0 && h.level <= level) {
+        // We've reached the end of the section...
+        break;
+      }
+      // Collapse/expand a sub-cell by setting the its `hidden` state:
+      w.setHidden(state);
+      continue;
+    }
+    // Code cells are more involved, as we need to analyze the outputs to check for generated Markdown/HTML...
+    const c = w as CodeCell;
+    const model = w.model as CodeCellModel;
+    const outputs = model.outputs;
+
+    // First, we do an initial pass to check for generated Markdown/HTML. If we don't find Markdown/HTML, then the entire cell can be collapsed/expanded (both inputs and outputs)...
+    let FLG = false;
+    let dtypes: string[] = [];
+    for (let j = 0; j < outputs.length; j++) {
+      // Retrieve the cell output model:
+      const m = outputs.get(j);
+
+      // Check the cell output MIME types:
+      dtypes = Object.keys(m.data);
+      for (let k = 0; k < dtypes.length; k++) {
+        const t = dtypes[k];
+        if (isMarkdown(t) || isDOM(t) || t === 'text/plain') {
+          // FIXME: apparent Jupyter bug where additional Markdown displays have a `text/plain` MIME type
```

Comment:
Ooh, interesting. Is there an issue for this? Can you provide a reproducible example?

Reply:
I did not find a related issue, but I cannot remember how hard I looked either. Basically, I discovered this in the course of generating multiple Markdown displays and logging `t`, noticing that the MIME type was not the expected Markdown MIME type.